src/ips.rs

```diff
@@ -251,28 +236,110 @@ impl Ips {
                 // importance to the network which can be risky if such node goes down.
                 candidates.sort_by(|a, b| {
-                    state.nodes[a.index]
+                    working_state.nodes[a.index]
                         .betweenness
-                        .partial_cmp(&state.nodes[b.index].betweenness)
+                        .partial_cmp(&working_state.nodes[b.index].betweenness)
                         .unwrap()
                 });
 
                 for peer in candidates.iter().take(peers_to_add_count as usize) {
                     peer_list_entry.list.push(peer.ip);
                 }
             }
+            working_state.peer_list.push(peer_list_entry);
+        }
+
+        // TODO(asmie): recalculate and compare factors to check if network is going better
+
+        working_state.peer_list
+    }
+
+    // Helper functions
 
-            // Do not compute factors one more time after every single peerlist addition. At least
-            // for now, when computing factors is very expensive (especially betweenness and closeness).
-            // Re-calculating it after each node for whole graph would take too long.
-            // TODO(asmie): recalculate some factors (like islands) after each node to check if graph is still connected
+    /// Generate peerlist for given nodes based on their connections
+    fn generate_peerlist(&self, nodes: &[Node]) -> Vec<Peer> {
+        let mut peer_list = Vec::with_capacity(nodes.len());
 
-            peer_list.push(peer_list_entry);
+        for node in nodes {
+            peer_list.push(self.generate_peerlist_for_node(node, nodes));
         }
+
         peer_list
     }
 
-    // Helper functions
+    /// Generate peerlist for given node based on its connections
+    fn generate_peerlist_for_node(&self, node: &Node, nodes: &[Node]) -> Peer {
+        let mut peer_list_entry = Peer {
+            ip: IpAddr::from_str(node.ip.as_str()).expect(ERR_PARSE_IP),
+            list: Vec::with_capacity(node.connections.len()),
+        };
+
+        for peer in &node.connections {
+            peer_list_entry
+                .list
+                .push(IpAddr::from_str(nodes[*peer].ip.as_str()).expect(ERR_PARSE_IP));
+        }
+
+        peer_list_entry
+    }
+
+    /// Generate state for IPS
+    fn generate_state(&self, nodes: &[Node]) -> IpsState {
+        let mut ips_state = IpsState {
+            nodes: nodes.to_vec(),
+            ..Default::default()
+        };
+
+        let mut graph = construct_graph(nodes);
+        let betweenness = graph.betweenness_centrality();
+        let closeness = graph.closeness_centrality();
+
+        // Recalculate factors with new graph
+        for node in ips_state.nodes.iter_mut() {
+            let ip = IpAddr::from_str(&node.ip).unwrap();
+            node.betweenness = *betweenness.get(&ip).unwrap_or(&0.0);
+            node.closeness = *closeness.get(&ip).unwrap_or(&0.0);
```

Comment:
If betweenness or closeness cannot be found for an IP address, wouldn't that be a bug, or a sign of some other pathological condition?  I'm not sure we should ignore that behaviour.
Or rather, can this happen if metrics have not yet been generated for a given node?

Reply:
Well, we're generating those factors for each node in graph just a couple of lines above so I cannot find any reason, why some nodes would not have those factors calculated. Therefore, you're right we can treat it like bug.

Changed to `expect`.